python/taichi/lang/impl.py

```diff
@@ -381,58 +382,45 @@ def get_runtime():
     return pytaichi
 
 
+def _check_in_range(npty, val):
+    iif = np.iinfo(npty)
+    if not iif.min <= val <= iif.max:
+        # This isn't the case we want to deal with: |val| does't fall into the valid range of either
+        # the signed or the unsigned type.
+        error(
+            f'Constant {val} has exceeded the range of {to_taichi_type(npty)}: [{iif.min}, {iif.max}]'
+        )
+
+
 def _clamp_unsigned_to_range(npty, val):
     # npty: np.int32 or np.int64
     iif = np.iinfo(npty)
     if iif.min <= val <= iif.max:
         return val
     cap = (1 << iif.bits)
-    if not 0 <= val < cap:
-        # We let pybind11 fail intentionally, because this isn't the case we want
-        # to deal with: |val| does't fall into the valid range of either
-        # the signed or the unsigned type.
-        return val
+    assert 0 <= val < cap
     new_val = val - cap
-    warn(
-        f'Constant {val} has exceeded the range of {iif.bits} int, clamped to {new_val}'
-    )
     return new_val
 
 
 @taichi_scope
 def make_constant_expr_i32(val):
     assert isinstance(val, (int, np.integer))
-    return Expr(
-        _ti_core.make_const_expr_i32(_clamp_unsigned_to_range(np.int32, val)))
+    return Expr(_ti_core.make_const_expr_ip(i32, val))
 
 
 @taichi_scope
-def make_constant_expr(val):
+def make_constant_expr(val, dtype):
     if isinstance(val, (int, np.integer)):
-        if pytaichi.default_ip in {i32, u32}:
-            # It is not always correct to do such clamp without the type info on
-            # the LHS, but at least this makes assigning constant to unsigned
-            # int work. See https://github.com/taichi-dev/taichi/issues/2060
-            return Expr(
-                _ti_core.make_const_expr_i32(
-                    _clamp_unsigned_to_range(np.int32, val)))
-        if pytaichi.default_ip in {i64, u64}:
-            return Expr(
-                _ti_core.make_const_expr_i64(
-                    _clamp_unsigned_to_range(np.int64, val)))
-        assert False
-    elif isinstance(val, (float, np.floating, np.ndarray)):
-        if pytaichi.default_fp == f32:
-            return Expr(_ti_core.make_const_expr_f32(val))
-        if pytaichi.default_fp == f64:
-            return Expr(_ti_core.make_const_expr_f64(val))
-        if pytaichi.default_fp == f16:
-            # Use f32 to interact with python
-            return Expr(_ti_core.make_const_expr_f32(val))
-        assert False
-    else:
-        raise TaichiTypeError(
-            f'Invalid constant scalar data type: {type(val)}')
+        constant_dtype = pytaichi.default_ip if dtype is None else dtype
+        _check_in_range(to_numpy_type(constant_dtype), val)
+        return Expr(
+            _ti_core.make_const_expr_ip(
+                constant_dtype, _clamp_unsigned_to_range(np.int64, val)))
+    if isinstance(val, (float, np.floating, np.ndarray)):
```

Comment:
IIUC, if we remove the try-except block in `Expr.__init__`, we can remove `np.ndarray` here as well.